prog/dftb+/prg_dftb/main.F90

```diff
@@ -0,0 +1,5637 @@
+      potential%intShell(:,:,1) = potential%intShell(:,:,1) + shellPot(:,:,1)
+    end if
+
+    if (allocated(spinW)) then
+      call getSpinShift(shellPot, chargePerShell, species, orb, spinW)
+      potential%intShell = potential%intShell + shellPot
+    end if
+
+    call total_shift(potential%intShell, potential%intAtom, orb, species)
+    call total_shift(potential%intBlock, potential%intShell, orb, species)
+
+  end subroutine addChargePotentials
+
+
+  !> Add potentials comming from on-site block of the dual density matrix.
+  subroutine addBlockChargePotentials(qBlockIn, qiBlockIn, tDftbU, tImHam, species, orb,&
+      & nDftbUFunc, UJ, nUJ, iUJ, niUJ, potential)
+
+    !> block input charges
+    real(dp), allocatable, intent(in) :: qBlockIn(:,:,:,:)
+
+    !> imaginary part
+    real(dp), allocatable, intent(in) :: qiBlockIn(:,:,:,:)
+
+    !> is this a +U calculation
+    logical, intent(in) :: tDftbU
+
+    !> does the hamitonian have an imaginary part in real space?
+    logical, intent(in) :: tImHam
+
+    !> chemical species of all atoms
+    integer, intent(in) :: species(:)
+
+    !> Orbital information
+    type(TOrbitals), intent(in) :: orb
+
+    !> choice of +U functional
+    integer, intent(in) :: nDftbUFunc
+
+    !> prefactor for +U potential
+    real(dp), allocatable, intent(in) :: UJ(:,:)
+
+    !> Number DFTB+U blocks of shells for each atom type
+    integer, intent(in), optional :: nUJ(:)
+
+    !> which shells are in each DFTB+U block
+    integer, intent(in), optional :: iUJ(:,:,:)
+
+    !> Number of shells in each DFTB+U block
+    integer, intent(in), optional :: niUJ(:,:)
+
+    !> potentials acting in system
+    type(TPotentials), intent(inout) :: potential
+
+
+    if (tDFTBU) then
+      if (tImHam) then
+        call getDftbUShift(potential%orbitalBlock, potential%iorbitalBlock, qBlockIn, qiBlockIn,&
+            & species,orb, nDFTBUfunc, UJ, nUJ, niUJ, iUJ)
+      else
+        call getDftbUShift(potential%orbitalBlock, qBlockIn, species, orb, nDFTBUfunc, UJ, nUJ,&
+            & niUJ, iUJ)
+      end if
+      potential%intBlock = potential%intBlock + potential%orbitalBlock
+    end if
+
+  end subroutine addBlockChargePotentials
+
+
+
+  !> Returns the Hamiltonian for the given scc iteration
+  subroutine getSccHamiltonian(H0, over, nNeighbor, neighborList, species, orb, iPair,&
+      & img2CentCell, potential, ham, iHam)
+
+    !> non-SCC hamitonian (sparse)
+    real(dp), intent(in) :: H0(:)
+
+    !> overlap (sparse)
+    real(dp), intent(in) :: over(:)
+
+    !> Number of atomic neighbours
+    integer, intent(in) :: nNeighbor(:)
+
+    !> list of atomic neighbours
+    type(TNeighborList), intent(in) :: neighborList
+
+    !> species of atoms
+    integer, intent(in) :: species(:)
+
+    !> atomic orbital information
+    type(TOrbitals), intent(in) :: orb
+
+    !> Index for atomic blocks in sparse data
+    integer, intent(in) :: iPair(:,:)
+
+
```

Comment:
I've removed superfluous argument. As for passing in `nAtom` instead of array sizing: I wanted to reduce the number of arguments, so I decided to deduce `nAtom` from array sizes in all routines. This is not the most elegant, I agree. Once we have a proper geometry container with coords, species, etc. inside, it could also have a field `nAtom`, and we won't need this hack... :wink: